### PR TITLE
Set up initial Django app.

### PR DESCRIPTION
**Pull request by nworden**
This sets up a Django app to run alongside the webapp2 app, with one
page as an example (the admin/statistics page, selected for being maybe
the simplest page we have).

Of course the main goal is to start the migration off webapp2 and onto 
Django. There's a couple other things going on too though:
- I put the new files in new places. There's been an open issue about this 
for a while (#212), and I figured that since I was going to need to create 
new references to these files anyway it would be a convenient time to put 
them in subdirectories. Did the same with the new test files.
- I started refactoring the auto-params list. Currently, the base webapp2 
handler has a giant list of parameters to read, which is shared across all 
handlers. The Django views are set up so that each view class has its own 
parameters; they call a shared function to read them, but the parameter 
list is not shared. If there end up being some views that share a lot of 
parameters, it shouldn't be a big deal: we can just put the shared 
parameter list somewhere shared and have the view classes call read_params 
an extra time with the shared parameter list.
- I set up a site_settings module. We have a bunch of stuff specific to 
Google hard-coded in various places, which would be nice to get away from 
doing, but we don't have a firm plan on how to address it (the general 
problem is that it's unclear where we want to be on the spectrum of "easy 
for a non-tech-savvy user to set up" vs. efficient). Putting the path 
prefix constant in site_settings.py is basically a way to punt on that 
question: whatever we want to do, we can do it later just by modifying 
site_settings (e.g., we could require users to set their own constants in 
the file when they set up their App Engine instance, or pass them through 
from app.yaml as environment variables, or use a config setting), and, no 
matter what we do, there's a good chance it'll be easier (and at least 
won't be harder) if the site-specific stuff is consolidated in one place.
- `server_test_cases/admin_statistics_tests.py` was simultaneously testing 
the UsageCounters and the /admin/statistics page, so I split it up.

### CODE REVIEW

**Comment by nworden** [comment]
The test failure is issue #226 again. It only failed on one of the Python 2.7 runs, the other fully passed.

File: app/views/base.py [hunk 1]
```diff
@@ -0,0 +1,216 @@
...
+        return '/%s%s' % (prefix, path)
+
+    def build_absolute_uri(self, path=None):
+        """Builds an absolute URI given a path.
+
+        See build_absolute_path (above) for an explanation of why we implement
+        this function ourselves.
+
+        Args:
+            path (str, optional): A path, e.g., 'abc?x=y'. Must NOT have a
```
Comment (gimite) [inline app/views/base.py]: Document its behavior when path is omitted.

File: app/views/base.py [hunk 2]
```diff
@@ -0,0 +1,216 @@
...
+
+    def render(self, template_name, **template_vars):
+        """Renders a template with the given variables.
+
+        Args:
+            template_name (str): The filename of the template in app/resources.
+            **template_vars: Named variables to pass to the template.
+
+        Returns:
+            HttpResponse: A HttpResponse with the rendered template.
```
Comment (gimite) [inline app/views/base.py]: Nit: A -> An

File: app/views/base.py [hunk 3]
```diff
@@ -0,0 +1,216 @@
...
+    def as_view(cls, **initkwargs):
+        # pylint: disable=E,W,R,C
+        # We want to have a setup() function called before dispatch() (see
+        # explanation in the comments on the setup() function), and Django 2.2
+        # has support for it built in. However, we're not on Django 2.2 yet, so
+        # we make the modification to View.as_view() ourselves to get this
+        # feature ahead of time (the code below is just a copy of the original
+        # Django 1.11 View.as_view function, with the setup() call stuck in). We
+        # can clean this up when we upgrade to Django 2.2.
+        """
```
Comment (gimite) [inline app/views/base.py]: Nit: Should be formatted in 1 (or 2) line:
```
"""Main entry point for a request-response process."""
```

File: app/wsgi.py [hunk 4]
```diff
@@ -0,0 +1,13 @@
...
+import os
+import sys
+
+from django.core import wsgi
+
+sys.path.append(os.path.join(os.path.dirname(__file__), 'vendors'))
+
+os.environ.setdefault('DJANGO_SETTINGS_MODULE', 'settings')
+
+# pylint: disable=invalid-name
```
Comment (gimite) [inline app/wsgi.py]: Just curious: Why is this needed?

File: tests/views/django_tests_base.py [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+        Args:
+            response (Response): A response from a Django test client.
+
+        Returns:
+            scrape.Document: A wrapper around the response's contents to help
+                with examining it.
+        """
+        # TODO(nworden): when everything's on Django, make some changes to
+        # scrape.py so it better fits Django's test framework.
+        return scrape.Document(response.content, None, response.status_code,
```
Comment (gimite) [inline tests/views/django_tests_base.py]: Optional: May be better to explicitly write param names at least for `None`s e.g., `url=None`.

File: tests/views/test_admin_statistics.py [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+class AdminStatisticsViewTests(django_tests_base.DjangoTestsBase):
+
+    def setUp(self):
+        super(AdminStatisticsViewTests, self).setUp()
+        model.Repo(key_name='haiti').put()
+        self.counter = model.UsageCounter.create('haiti')
+        self.login(is_admin=True)
+
+    def test_person_counter(self):
+        setattr(self.counter, 'person', 3)
```
Comment (gimite) [inline tests/views/test_admin_statistics.py]: Just curious: Can't you just write `self.counter.person = 3`?

File: app/views/base.py [hunk 7]
```diff
@@ -0,0 +1,216 @@
...
+        # TODO(nworden): don't forget to call super.setup here once we upgrade
+        # to Django 2.2.
+        del request, args  # unused
+
+        # Set up the parameters and read in the base set of parameters.
+        self.params = utils.Struct()
+        self.read_params(get_params=BaseView._GET_PARAMETERS)
+
+        # Set up env variable with data needed by the whole app.
+        self.env = utils.Struct()
```
[the remark this answers is not in the document]
Comment (nworden) [reply app/views/base.py]: I played around with this for a while, but it's trickier than I expected, and I can't come up with a way to do it that doesn't feel over-engineered and/or too inflexible. In particular, a nice thing about Structs is it's no problem for subclass views to add their own data (and we could define utility functions to add data to them too if we wanted), but that gets complicated when you try to make it a class.
The seemingly obvious solution is to have the Env class be subclassed alongside the View subclasses, and I tried implementing it that way (I pushed it on a separate branch so you can see the comparison [here](https://github.com/nworden/personfinder/compare/djangoapp...nworden:djangoapp-envclass)). It does work, but I feel like doing it this way would cause us to lose a lot of flexibility. For example, if a subclass needed to define an env variable based on one of its own parameters (like the base class does with the lang parameter), it couldn't do it, because its params haven't been read in yet.
Do you have a better idea of how to implement it as a class?

File: app/wsgi.py (hunk 4 above)
Comment (gimite) [reply app/wsgi.py]: The naming sounds good, but can you comment the reason of this lint exemption if you stick with it?

File: app/manage.py [hunk 8]
```diff
@@ -0,0 +1,22 @@
+import os
```
[the remark this answers is not in the document]
Comment (nworden) [reply app/manage.py]: I guess at this point the auto-generated stuff is gone, so agreed, added the copyright.

File: app/site_settings.py [hunk 9]
```diff
@@ -11,4 +25,15 @@
...
 
-OPTIONAL_PATH_PREFIX = 'personfinder/'
+# When Person Finder is run at an address under a subdirectory rather than at
+# root (e.g., www.organization.org/personfinder/), that subdirectory should be
+# specified here so that URLs can be handled correctly. The value MUST NOT start
+# or end with a slash.
+OPTIONAL_PATH_PREFIX = 'personfinder'
+
+# Hosts allowed in prod (not applicable to local servers).
+PROD_ALLOWED_HOSTS = [
```
[the remark this answers is not in the document]
Comment (nworden) [reply app/site_settings.py]: It's a security measure ([here's the blog entry where they explained it](https://www.djangoproject.com/weblog/2013/feb/19/security/#s-issue-host-header-poisoning), and there's some more detail in [the settings docs](https://docs.djangoproject.com/en/1.11/ref/settings/#allowed-hosts)). The setting is required, so we have to put something there. I've set it to a wildcard for now (seems ok since afaik we/webapp2 weren't doing any host header validation before), but I put in a TODO to try to set it for real later.

File: app/urls.py [hunk 10]
```diff
@@ -1,22 +1,36 @@
...
+
 """URL routing module."""
 
 from django.conf import urls
 
 import site_settings
 import views.admin.statistics
 
-_BASE_URL_PATTERNS = [('global/admin/statistics/',
+_BASE_URL_PATTERNS = [('global/admin/statistics/?',
```
[the remark this answers is not in the document]
Comment (nworden) [reply app/urls.py]: Done.

[the remark this answers is not in the document]
Comment (nworden) [reply app/urls.py]: Done.

File: app/urls.py [hunk 11]
```diff
@@ -0,0 +1,22 @@
...
+
+from django.conf import urls
+
+import site_settings
+import views.admin.statistics
+
+_BASE_URL_PATTERNS = [('global/admin/statistics/',
+                       views.admin.statistics.AdminStatisticsView.as_view)]
+
+# pylint: disable=invalid-name
```
[the remark this answers is not in the document]
Comment (nworden) [reply app/urls.py]: Done.

File: app/urls.py [hunk 12]
```diff
@@ -1,22 +1,36 @@
...
 
-_BASE_URL_PATTERNS = [('global/admin/statistics/',
+_BASE_URL_PATTERNS = [('global/admin/statistics/?',
                        views.admin.statistics.AdminStatisticsView.as_view)]
 
 # pylint: disable=invalid-name
 urlpatterns = [
-    urls.url('^%s' % base_pattern[0], base_pattern[1]())
-    for base_pattern in _BASE_URL_PATTERNS
+    urls.url('^%s' % path_exp, view_func())
```
[the remark this answers is not in the document]
Comment (nworden) [reply app/urls.py]: Fixed.

File: app/views/base.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (nworden) [reply app/views/base.py]: > Do you know how many of views define its own env fields? IMHO env should provide general information about the environment, so it shouldn't be customized for every view. Having BaseViewEnv and AdminBaseViewEnv sounds good, but maybe there should be only these two?

I think when we get to end-user-facing views, the parent class for them will need some settings of its own too (e.g., to set whether it's lite mode).

> Can you elaborate that the problem is? Since looks like you are able to access param with no problem here

The base-level Env class can access params, but if the AdminEnv class needed to access admin-specific params, it couldn't, because the AdminView setup function wouldn't have called read_params yet. That didn't matter for the admin views, but it'll be an issue for stuff like the `ui` param for end-user views. Also, more generally, I'm a little bit nervous that we might end up with other things where it'd be easier to initialize the env at the same time we're initializing the view to avoid duplicated code. So far I've only needed a part of what main.py's doing.

If the Env classes didn't have logic, and the field-setting was still left to View.setup, I wouldn't be concerned about this though. We could set up Env classes/subclasses that only allowed specific pre-named properties to avoid typo risks and make it easier to see what env objects contain, but still have their fields set by the View itself. What do you think?

In any case, I would prefer to handle all this separately and stick with Struct envs for this PR.

File: app/views/base.py [hunk 13]
```diff
@@ -0,0 +1,216 @@
...
+            get_params (dict): A dictionary from GET parameter keys to validator
+                functions.
+            post_params (dict): A dictionary from POST parameter keys to
+                validator functions.
+            file_params (dict): A dictionary from FILES parameter keys to
+                validator functions.
+        """
+        if self.request.method == 'GET':
+            if get_params:
+                for key, validator in get_params.items():
```
[the remark this answers is not in the document]
Comment (nworden) [reply app/views/base.py]: Probably the six library has something we could use to deal with it if we have to.

File: app/views/base.py [hunk 14]
```diff
@@ -102,29 +102,54 @@ def read_params(self, get_params=None, post_params=None, file_params=None):
...
         use.
 
         Args:
-            path (str, optional): A path, e.g., 'abc?x=y'. Must NOT have a
-                leading slash. May include a query string. If not specified,
-                will return the current path.
+            path (str, optional): A path (may include a query string), e.g.,
+                'abc?x=y'. Paths beginning with a slash will be interpreted as
+                being relative to the Django app's directory, while paths
+                without a leading slash will be considered relative to the
```
[the remark this answers is not in the document]
Comment (nworden) [reply app/views/base.py]: Yeah, you're right, I overcomplicated it. Changed it to just paths relative to the root directory.

File: app/wsgi.py (hunk 4 above)
Comment (nworden) [reply app/wsgi.py]: Done.

File: app/site_settings.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (gimite) [reply app/site_settings.py]: I see, thanks for the explanation!

File: app/views/base.py [hunk 15]
```diff
@@ -119,37 +133,23 @@ def build_absolute_path(self, path=None):
...
-                    if self._request_is_for_prefixed_path()
-                    else path)
-        else:
-            # We're dealing with a relative path, so just append it to the
-            # current request's path (which will already include the path prefix
-            # if it's being used).
-            return ('%s%s' % (self.request.path, path)
-                    if self.request.path[-1] == '/'
-                    else '%s/%s' % (self.request.path, path))
+        return ('/%s%s' % (site_settings.OPTIONAL_PATH_PREFIX, path)
```
Comment (gimite) [inline app/views/base.py]: Can you add an assertion that it starts with "/"?

Comment (gimite) [inline app/views/base.py]: Optional: This statement may be more readable by formatting as:
```python
if ...:
    return ...
else:
    return ...
```

File: app/views/base.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (gimite) [reply app/views/base.py]: > > Do you know how many of views define its own env fields? IMHO env should provide general information about the environment, so it shouldn't be customized for every view. Having BaseViewEnv and AdminBaseViewEnv sounds good, but maybe there should be only these two?
> 
> I think when we get to end-user-facing views, the parent class for them will need some settings of its own too (e.g., to set whether it's lite mode).

I see, then there can be more than 2, but the number is still limited (for each base class like AppBase, StaticBase, but not for every view), right?

> 
> > Can you elaborate that the problem is? Since looks like you are able to access param with no problem here
> 
> The base-level Env class can access params, but if the AdminEnv class needed to access admin-specific params, it couldn't, because the AdminView setup function wouldn't have called read_params yet. 

Oh I see. This particular issue can be workarounded if you make either `param` or `__GET_PARAMETERS` a method (instead of a variable), and let subclasses override it. This may be a good thing on its own because you can avoid an issue that the value of `self.params` is different depending on when you access it.

> That didn't matter for the admin views, but it'll be an issue for stuff like the `ui` param for end-user views. Also, more generally, I'm a little bit nervous that we might end up with other things where it'd be easier to initialize the env at the same time we're initializing the view to avoid duplicated code. So far I've only needed a part of what main.py's doing.

I agree that it can potentially complicate the initialization problem than it currently is.

> 
> If the Env classes didn't have logic, and the field-setting was still left to View.setup, I wouldn't be concerned about this though. We could set up Env classes/subclasses that only allowed specific pre-named properties to avoid typo risks and make it easier to see what env objects contain, but still have their fields set by the View itself. What do you think?

That sounds like a good idea. One reason I didn't like `env` to be a struct is that it is hard to know what fields it has. Its class definition would serve as a useful documentation of the fields.

> 
> In any case, I would prefer to handle all this separately and stick with Struct envs for this PR.

SGTM.

File: app/views/base.py (hunk 15 above)
Comment (nworden) [reply app/views/base.py]: Done.

Comment (nworden) [reply app/views/base.py]: Done.

**Comment by nworden** [comment]
The test failure is more of issue #226. The 2.7.12 tests passed.

File: app/views/base.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (nworden) [reply app/views/base.py]: > I see, then there can be more than 2, but the number is still limited (for each base class like AppBase, StaticBase, but not for every view), right?

Right.

> One reason I didn't like env to be a struct...

Are there other reasons (especially any that my proposal wouldn't address)?

[the remark this answers is not in the document]
Comment (gimite) [reply app/views/base.py]: Another reason is that, in our original (webapp2) implementation, there's lengthy code which initializes `env` directly written in main.py:
https://github.com/google/personfinder/blob/master/app/main.py#L261-L445
which I wasn't quite happy with.

It was partially already resolved in the new implementation because it's in a better place and is shorter (at least for now). So that would be OK. If it gets longer again, we may factor it out into a utility function somewhere.
